File: src/main/java/mods/eln/sixnode/wirelesssignal/WirelessUtils.java
package mods.eln.sixnode.wirelesssignal;

import mods.eln.misc.Coordinate;
import mods.eln.sixnode.wirelesssignal.tx.WirelessSignalTxElement;
import net.minecraft.block.Block;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;

import java.util.*;
import java.util.Map.Entry;

public class WirelessUtils {

    private WirelessUtils() {
    }

    public static void getTx(IWirelessSignalSpot root, HashMap<String, HashSet<IWirelessSignalTx>> txSet, HashMap<IWirelessSignalTx, Double> txStrength) {
        HashSet<IWirelessSignalSpot> spotSet = new HashSet<IWirelessSignalSpot>();
        txSet.clear();
        txStrength.clear();
        getTx(root, txSet, txStrength, spotSet, true, 0);
    }

    private static void getTx(IWirelessSignalSpot from, HashMap<String, HashSet<IWirelessSignalTx>> txSet, HashMap<IWirelessSignalTx, Double> txStrength, HashSet<IWirelessSignalSpot> spotSet, boolean isRoot, double strength) {
        if (spotSet.contains(from)) return;

        spotSet.add(from);

        if (!isRoot) {
            for (List<IWirelessSignalTx> txs : from.getTx().values()) {
                for (IWirelessSignalTx tx : txs) {
                    if (isRoot)
                        strength = tx.getRange() - getVirtualDistance(tx.getCoordinate(), from.getCoordinate(), tx.getCoordinate().trueDistanceTo(from.getCoordinate()));
                    addTo(tx, strength, txSet, txStrength);
                }
            }
            for (IWirelessSignalSpot spot : from.getSpot()) {
                if (isRoot)
                    strength = spot.getRange() - getVirtualDistance(spot.getCoordinate(), from.getCoordinate(), spot.getCoordinate().trueDistanceTo(from.getCoordinate()));
                getTx(spot, txSet, txStrength, spotSet, false, strength);
            }
        } else {
            LinkedList<IWirelessSignalSpot> spots = new LinkedList<IWirelessSignalSpot>();
            spots.addAll(from.getSpot());

            LinkedList<IWirelessSignalTx> txs = new LinkedList<IWirelessSignalTx>();
            for (ArrayList<IWirelessSignalTx> txss : from.getTx().values()) {
                txs.addAll(txss);
            }

            double bestScore;
            Object best = null;
            while (!spots.isEmpty() || !txs.isEmpty()) {
                bestScore = Double.MAX_VALUE;
                for (IWirelessSignalSpot spot : spots) {
                    double temp = spot.getCoordinate().trueDistanceTo(from.getCoordinate());
                    if (temp < bestScore) {
                        bestScore = temp;
                        best = spot;
                    }
                }

                for (IWirelessSignalTx tx : txs) {
                    double temp = tx.getCoordinate().trueDistanceTo(from.getCoordinate());
                    if (temp < bestScore) {
                        bestScore = temp;
                        best = tx;
                    }
                }

                if (best instanceof IWirelessSignalSpot) {
                    IWirelessSignalSpot b = (IWirelessSignalSpot) best;
                    if (isRoot)
                        strength = b.getRange() - getVirtualDistance(b.getCoordinate(), from.getCoordinate(), b.getCoordinate().trueDistanceTo(from.getCoordinate()));
                    getTx(b, txSet, txStrength, spotSet, false, strength);
                    spots.remove(best);
                } else if (best == null) {
                    break;
                } else {
                    IWirelessSignalTx tx = (IWirelessSignalTx) best;

                    if (isRoot)
                        strength = tx.getRange() - getVirtualDistance(tx.getCoordinate(), from.getCoordinate(), tx.getCoordinate().trueDistanceTo(from.getCoordinate()));
                    addTo(tx, strength, txSet, txStrength);
                    txs.remove(best);
                }
            }
        }
    }

    private static void addTo(IWirelessSignalTx tx, double strength, Map<String, HashSet<IWirelessSignalTx>> reg, Map<IWirelessSignalTx, Double> txStrength) {
        String channel = tx.getChannel();
        HashSet<IWirelessSignalTx> ch = reg.get(channel);
        if (ch != null && ch.contains(tx)) return;
        if (ch == null)
            reg.put(channel, ch = new HashSet<IWirelessSignalTx>());
        ch.add(tx);
        txStrength.put(tx, strength);
    }

	/*
     *
	 * public static HashSet<IWirelessSignalTx> getTx(String channel,IWirelessSignalSpot root){ HashSet<IWirelessSignalTx> txSet = new HashSet<IWirelessSignalTx>(); getTx(channel, root,txSet); return txSet; }
	 * 
	 * private static void getTx(String channel,IWirelessSignalSpot root,HashSet<IWirelessSignalTx> txSet){ for(IWirelessSignalSpot spot : root.getSpot()){ getTx(channel, spot, txSet); }
	 * 
	 * if(channel != null){ ArrayList<IWirelessSignalTx> txs = root.getTx().get(channel); if(txs != null) txSet.addAll(txs); }else{ for(ArrayList<IWirelessSignalTx> txs : root.getTx().values()){ txSet.addAll(txs); } } }
	 */

    public static WirelessSignalSpot buildSpot(Coordinate c, String channel, int range) {
        HashMap<String, ArrayList<IWirelessSignalTx>> txs = new HashMap<String, ArrayList<IWirelessSignalTx>>();
        ArrayList<IWirelessSignalSpot> spots = new ArrayList<IWirelessSignalSpot>();

        for (IWirelessSignalSpot spot : IWirelessSignalSpot.spots) {
            if (isInRange(spot.getCoordinate(), c, spot.getRange())) {
                spots.add(spot);
            }
        }

        if (channel != null) {
            ArrayList<IWirelessSignalTx> inRangeTx = new ArrayList<IWirelessSignalTx>();

            List<IWirelessSignalTx> sameChannelTx = WirelessSignalTxElement.channelMap.get(channel);
            if (sameChannelTx != null) {
                for (IWirelessSignalTx tx : sameChannelTx) {
                    if (isInRange(tx.getCoordinate(), c, tx.getRange())) {
                        inRangeTx.add(tx);
                    }
                }
            }
            if (!inRangeTx.isEmpty())
                txs.put(channel, inRangeTx);
        } else {
            for (Entry<String, ArrayList<IWirelessSignalTx>> entryTxs : WirelessSignalTxElement.channelMap.entrySet()) {
                ArrayList<IWirelessSignalTx> inRangeTx = new ArrayList<IWirelessSignalTx>();

                for (IWirelessSignalTx tx : entryTxs.getValue()) {
                    if (isInRange(tx.getCoordinate(), c, tx.getRange())) {
                        inRangeTx.add(tx);
                    }
                }

                if (!inRangeTx.isEmpty())
                    txs.put(entryTxs.getKey(), inRangeTx);
            }
        }

        return new WirelessSignalSpot(txs, spots, c, range);
    }

    static private boolean isInRange(Coordinate txC, Coordinate rxC, double range) {
        double distance = txC.trueDistanceTo(rxC);
        if (distance > range) return false;
        if (getVirtualDistance(txC, rxC, distance) > range) return false;
        return true;
    }

    static private double getVirtualDistance(Coordinate txC, Coordinate rxC, double distance) {
        double virtualDistance = distance;
        if (distance > 2) {
            double vx, vy, vz;
            double dx, dy, dz;
            vx = rxC.pos.getX() + 0.5;
            vy = rxC.pos.getY() + 0.5;
            vz = rxC.pos.getZ() + 0.5;

            dx = (txC.pos.getX() - rxC.pos.getX()) / distance;
            dy = (txC.pos.getY() - rxC.pos.getY()) / distance;
            dz = (txC.pos.getZ() - rxC.pos.getZ()) / distance;
            Coordinate c = new Coordinate();
            c.setDimension(rxC.getDimension());

            for (int idx = 0; idx < distance - 1; idx++) {
                vx += dx;
                vy += dy;
                vz += dz;
                c.setPosition(new Vec3d(vx, vy, vz));
                if (c.doesBlockExist()) {
                    Block b = c.getBlockState().getBlock();
                    World w = c.world();

                    virtualDistance +=
                        b.isOpaqueCube(b.getDefaultState()) &&
                        !b.isAir(b.getDefaultState(), w,new BlockPos(c.pos.getX(), c.pos.getY(), c.pos.getZ())) ?
                        2.0 : 0.0;
                }
            }
        }
        return virtualDistance;
    }

    public static class WirelessSignalSpot implements IWirelessSignalSpot {

        HashMap<String, ArrayList<IWirelessSignalTx>> txs;
        ArrayList<IWirelessSignalSpot> spots;
        Coordinate coordinate;
        int range;

        public WirelessSignalSpot(HashMap<String, ArrayList<IWirelessSignalTx>> txs, ArrayList<IWirelessSignalSpot> spots, Coordinate coordinate, int range) {
            this.txs = txs;
            this.spots = spots;
            this.coordinate = coordinate;
            this.range = range;
        }

        @Override
        public HashMap<String, ArrayList<IWirelessSignalTx>> getTx() {
            return txs;
        }

        @Override
        public ArrayList<IWirelessSignalSpot> getSpot() {
            return spots;
        }

        @Override
        public Coordinate getCoordinate() {
            return coordinate;
        }

        @Override
        public int getRange() {
            return range;
        }
    }
}

File: src/main/java/mods/eln/sixnode/wirelesssignal/WirelessUtils.kt
package mods.eln.sixnode.wirelesssignal

import mods.eln.misc.Coordonate
import mods.eln.misc.Line3f
import mods.eln.misc.Utils
import mods.eln.misc.Vec3f
import mods.eln.sixnode.wirelesssignal.tx.WirelessSignalTxElement
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.init.Blocks
import net.minecraft.server.MinecraftServer
import net.minecraft.world.World
import net.minecraft.world.chunk.Chunk

import java.util.*
import kotlin.collections.Map.Entry

data class WirelessRaytraceKey(val startCoordonate: Coordonate, val endCoordonate: Coordonate)

fun costOfBlockContaining(seg: Line3f, dim: Int) = seg.length() * when(seg.midpoint().toCoordonate(dim).block.material) {
    Material.air -> 1.0
    else -> 2.0
}

class WirelessCacheData(var segments: List<Line3f>, var dim: Int) {
    var validTick: Int = 0
    var cachedCost: Double = 0.0
    var costMap: MutableMap<Chunk, Double> = HashMap()
    var chunkMap: MutableMap<Chunk, MutableList<Line3f>> = HashMap()

    init {
        for(seg in segments) {
            chunkMap.getOrPut(seg.midpoint().toCoordonate(dim).chunk, { -> ArrayList<Line3f>() }).add(seg)
        }
        populateCosts(true)
    }

    fun populateCosts(full: Boolean = false): Double {
        var curTick = MinecraftServer.getServer().tickCounter
        if(validTick == curTick)
            return cachedCost
        validTick = curTick

        var total: Double = 0.0

        for((chunk, segments) in chunkMap) {
            if(full || chunk.isChunkLoaded) {
                var chunk_total = 0.0
                for(seg in segments) {
                    chunk_total += costOfBlockContaining(seg, dim)
                }
                costMap.put(chunk, chunk_total)
            }
            total += costMap.getOrDefault(chunk, 0.0)
        }

        cachedCost = total
        return total
    }
}


object WirelessUtils {

    fun getTx(root: IWirelessSignalSpot, txSet: HashMap<String, HashSet<IWirelessSignalTx>>, txStrength: HashMap<IWirelessSignalTx, Double>) {
        val spotSet = HashSet<IWirelessSignalSpot>()
        txSet.clear()
        txStrength.clear()
        getTx(root, txSet, txStrength, spotSet, true, 0.0)
    }

    private fun getTx(from: IWirelessSignalSpot, txSet: HashMap<String, HashSet<IWirelessSignalTx>>, txStrength: HashMap<IWirelessSignalTx, Double>, spotSet: HashSet<IWirelessSignalSpot>, isRoot: Boolean, strength: Double) {
        var strength = strength
        if (spotSet.contains(from)) return

        spotSet.add(from)

        if (!isRoot) {
            for (txs in from.tx.values) {
                for (tx in txs) {
                    if (isRoot)
                        strength = tx.range - getVirtualDistance(tx.coordonate, from.coordonate, tx.coordonate.trueDistanceTo(from.coordonate))
                    addTo(tx, strength, txSet, txStrength)
                }
            }
            for (spot in from.spot) {
                if (isRoot)
                    strength = spot.range - getVirtualDistance(spot.coordonate, from.coordonate, spot.coordonate.trueDistanceTo(from.coordonate))
                getTx(spot, txSet, txStrength, spotSet, false, strength)
            }
        } else {
            val spots = LinkedList<IWirelessSignalSpot>()
            spots.addAll(from.spot)

            val txs = LinkedList<IWirelessSignalTx>()
            for (txss in from.tx.values) {
                txs.addAll(txss)
            }

            var bestScore: Double
            var best: Any? = null
            while (!spots.isEmpty() || !txs.isEmpty()) {
                bestScore = java.lang.Double.MAX_VALUE
                for (spot in spots) {
                    val temp = spot.coordonate.trueDistanceTo(from.coordonate)
                    if (temp < bestScore) {
                        bestScore = temp
                        best = spot
                    }
                }

                for (tx in txs) {
                    val temp = tx.coordonate.trueDistanceTo(from.coordonate)
                    if (temp < bestScore) {
                        bestScore = temp
                        best = tx
                    }
                }

                if (best is IWirelessSignalSpot) {
                    val b = best
                    if (isRoot)
                        strength = b.range - getVirtualDistance(b.coordonate, from.coordonate, b.coordonate.trueDistanceTo(from.coordonate))
                    getTx(b, txSet, txStrength, spotSet, false, strength)
                    spots.remove(best)
                } else if (best == null) {
                    break
                } else if (best is IWirelessSignalTx){
                    val tx = best

                    if (isRoot)
                        strength = tx.range - getVirtualDistance(tx.coordonate, from.coordonate, tx.coordonate.trueDistanceTo(from.coordonate))
                    addTo(tx, strength, txSet, txStrength)
                    txs.remove(best)
                }else{
                    break
                }
            }
        }
    }

    private fun addTo(tx: IWirelessSignalTx, strength: Double, reg: MutableMap<String, HashSet<IWirelessSignalTx>>, txStrength: MutableMap<IWirelessSignalTx, Double>) {
        val channel = tx.channel
        var ch: HashSet<IWirelessSignalTx>? = reg[channel]
        if (ch != null && ch.contains(tx)) return
        if (ch == null)
            ch = HashSet<IWirelessSignalTx>()
            reg.put(channel, ch)
        ch.add(tx)
        txStrength.put(tx, strength)
    }

    /*
     *
	 * public static HashSet<IWirelessSignalTx> getTx(String channel,IWirelessSignalSpot root){ HashSet<IWirelessSignalTx> txSet = new HashSet<IWirelessSignalTx>(); getTx(channel, root,txSet); return txSet; }
	 *
	 * private static void getTx(String channel,IWirelessSignalSpot root,HashSet<IWirelessSignalTx> txSet){ for(IWirelessSignalSpot spot : root.getSpot()){ getTx(channel, spot, txSet); }
	 *
	 * if(channel != null){ ArrayList<IWirelessSignalTx> txs = root.getTx().get(channel); if(txs != null) txSet.addAll(txs); }else{ for(ArrayList<IWirelessSignalTx> txs : root.getTx().values()){ txSet.addAll(txs); } } }
	 */

    fun buildSpot(c: Coordonate, channel: String?, range: Int): WirelessSignalSpot {
        val txs = HashMap<String, ArrayList<IWirelessSignalTx>>()
        val spots = ArrayList<IWirelessSignalSpot>()

        for (spot in IWirelessSignalSpot.spots) {
            if (isInRange(spot.coordonate, c, spot.range.toDouble())) {
                spots.add(spot)
            }
        }

        if (channel != null) {
            val inRangeTx = ArrayList<IWirelessSignalTx>()

            val sameChannelTx = WirelessSignalTxElement.channelMap[channel]
            if (sameChannelTx != null) {
                for (tx in sameChannelTx) {
                    if (isInRange(tx.coordonate, c, tx.range.toDouble())) {
                        inRangeTx.add(tx)
                    }
                }
            }
            if (!inRangeTx.isEmpty())
                txs.put(channel, inRangeTx)
        } else {
            for ((key, value) in WirelessSignalTxElement.channelMap) {
                val inRangeTx = ArrayList<IWirelessSignalTx>()

                for (tx in value) {
                    if (isInRange(tx.coordonate, c, tx.range.toDouble())) {
                        inRangeTx.add(tx)
                    }
                }

                if (!inRangeTx.isEmpty())
                    txs.put(key, inRangeTx)
            }
        }

        return WirelessSignalSpot(txs, spots, c, range)
    }

    private fun isInRange(txC: Coordonate, rxC: Coordonate, range: Double): Boolean {
        val distance = txC.trueDistanceTo(rxC)
        if (distance > range) return false
        return getVirtualDistance(txC, rxC, distance) <= range
    }

    public val raytraceCache = Hashtable<WirelessRaytraceKey, WirelessCacheData>()

    private fun getDataForRaycast(txC: Coordonate, rxC: Coordonate): WirelessCacheData {
        val key = WirelessRaytraceKey(txC, rxC)

        var cache = raytraceCache[key]
        if(cache !== null) return cache

        val segments = Line3f.fromEndpoints(Vec3f.fromVolumeCenter(txC), Vec3f.fromVolumeCenter(rxC)).allSegmentsAlong()
        cache = WirelessCacheData(segments, txC.dimention)
        raytraceCache.put(key, cache)
        return cache
    }

    private fun getVirtualDistance(txC: Coordonate, rxC: Coordonate, distance: Double): Double {
        var cache = getDataForRaycast(txC, rxC)
        return cache.populateCosts()
    }

    class WirelessSignalSpot(internal var txs: HashMap<String, ArrayList<IWirelessSignalTx>>, internal var spots: ArrayList<IWirelessSignalSpot>, internal var coordonate: Coordonate, internal var range: Int) : IWirelessSignalSpot {

        override fun getTx(): HashMap<String, ArrayList<IWirelessSignalTx>> {
            return txs
        }

        override fun getSpot(): ArrayList<IWirelessSignalSpot> {
            return spots
        }

        override fun getCoordonate(): Coordonate {
            return coordonate
        }

        override fun getRange(): Int {
            return range
        }
    }
}
